File: c++/QEasyWidgets/Components/ToolTip.cpp
#include "ToolTip.h"

#include <QApplication>
#include <QScreen>
#include <QCursor>


ToolTipBase::ToolTipBase(QWidget *parent, const QString &text)
    : QFrame(parent)
    , m_text(text)
    , m_duration(333)
    , m_timer(nullptr)
    , m_container(nullptr)
    , m_containerLayout(nullptr)
    , m_label(nullptr)
    , m_opacityAnim(nullptr) {
    
    m_timer = new QTimer(this);
    m_timer->setSingleShot(true);
    connect(m_timer, &QTimer::timeout, this, &ToolTipBase::hide);
    
    // Set layout
    setLayout(new QHBoxLayout());
    layout()->setContentsMargins(12, 8, 12, 12);
    
    // Set container
    m_container = createContainer();
    m_containerLayout = new QHBoxLayout(m_container);
    layout()->addWidget(m_container);
    m_label = new QLabel(text, this);
    m_containerLayout->addWidget(m_label);
    m_containerLayout->setContentsMargins(8, 6, 8, 6);
    
    // Add shadow
    setDropShadowEffect(this, 21.0, QColor(0, 0, 0, 60), 0.0, 3.0);
    
    // Add opacity effect
    m_opacityAnim = setOpacityEffect(this);
    
    // Set style
    setAttribute(Qt::WA_TransparentForMouseEvents);
    setAttribute(Qt::WA_TranslucentBackground);
    setWindowFlags(Qt::Tool | Qt::FramelessWindowHint);
    setQSS();
}


QFrame* ToolTipBase::createContainer() {
    return new QFrame(this);
}


void ToolTipBase::setQSS() {
    m_container->setObjectName("container");
    m_label->setObjectName("contentLabel");
    StyleSheetBase::apply(this, StyleSheetBase::ToolTip);
    m_label->adjustSize();
    adjustSize();
}


void ToolTipBase::showEvent(QShowEvent *event) {
    m_opacityAnim->setStartValue(0);
    m_opacityAnim->setEndValue(1);
    m_opacityAnim->start();
    m_timer->stop();
    if (duration() > 0) {
        m_timer->start(m_duration + m_opacityAnim->duration());
    }
    QFrame::showEvent(event);
}


void ToolTipBase::hideEvent(QHideEvent *event) {
    m_timer->stop();
    QFrame::hideEvent(event);
}


QString ToolTipBase::text() const {
    return m_text;
}


void ToolTipBase::setText(const QString &text) {
    m_text = text;
    m_label->setText(text);
    m_container->adjustSize();
    adjustSize();
}


int ToolTipBase::duration() const {
    return m_duration;
}


void ToolTipBase::setDuration(int duration) {
    m_duration = duration;
}


void ToolTipBase::showText(const QPoint &pos, const QString &text) {
    move(pos);
    setText(text);
    show();
}


void ToolTipBase::showText(int x, int y, const QString &text) {
    move(x, y);
    setText(text);
    show();
}


void ToolTipBase::hideText() {
    hide();
}


void ToolTipBase::adjustPos(QWidget *widget, Position position) {
    QPoint pos;
    int x = 0, y = 0;
    
    switch (position) {
        case Position::Top:
            pos = widget->mapToGlobal(QPoint());
            x = pos.x() + widget->width() / 2 - width() / 2;
            y = pos.y() - height();
            break;
        case Position::Bottom:
            pos = widget->mapToGlobal(QPoint());
            x = pos.x() + widget->width() / 2 - width() / 2;
            y = pos.y() + widget->height();
            break;
        case Position::Left:
            pos = widget->mapToGlobal(QPoint());
            x = pos.x() - width();
            y = pos.y() + (widget->height() - height()) / 2;
            break;
        case Position::Right:
            pos = widget->mapToGlobal(QPoint());
            x = pos.x() + widget->width();
            y = pos.y() + (widget->height() - height()) / 2;
            break;
        case Position::TopLeft:
            pos = widget->mapToGlobal(QPoint());
            x = pos.x() - layout()->contentsMargins().left();
            y = pos.y() - height();
            break;
        case Position::TopRight:
            pos = widget->mapToGlobal(QPoint());
            x = pos.x() + widget->width() - width() + layout()->contentsMargins().right();
            y = pos.y() - height();
            break;
        case Position::BottomLeft:
            pos = widget->mapToGlobal(QPoint());
            x = pos.x() - layout()->contentsMargins().left();
            y = pos.y() + widget->height();
            break;
        case Position::BottomRight:
            pos = widget->mapToGlobal(QPoint());
            x = pos.x() + widget->width() - width() + layout()->contentsMargins().right();
            y = pos.y() + widget->height();
            break;
    }
    
    QRect screenRect = getScreenGeometry(getCurrentScreen());
    x = qMax(screenRect.left(), qMin(x, screenRect.right() - width() - 4));
    y = qMax(screenRect.top(), qMin(y, screenRect.bottom() - height() - 4));
    move(x, y);
}


// ToolTipEventFilter implementation
ToolTipEventFilter::ToolTipEventFilter(QWidget *parent, Position position, int delay)
    : QObject(parent)
    , m_showDelay(delay)
    , m_isEnter(false)
    , m_position(position)
    , m_timer(nullptr)
    , m_toolTip(nullptr) {
    init(parent, delay);
    m_position = position;
    m_toolTip = new ToolTipBase(parent->window(), parent->toolTip());
}


ToolTipEventFilter::ToolTipEventFilter(QWidget *parent, ToolTipBase *toolTip, int delay)
    : QObject(parent)
    , m_showDelay(delay)
    , m_isEnter(false)
    , m_timer(nullptr)
    , m_toolTip(nullptr) {
    init(parent, delay);
    m_toolTip = toolTip ? toolTip : new ToolTipBase(parent->window(), parent->toolTip());
}


void ToolTipEventFilter::init(QWidget *parent, int delay) {
    m_showDelay = delay;
    m_isEnter = false;
    
    m_timer = new QTimer(this);
    m_timer->setSingleShot(true);
    connect(m_timer, &QTimer::timeout, this, &ToolTipEventFilter::showToolTip);
}


bool ToolTipEventFilter::eventFilter(QObject *obj, QEvent *event) {
    Q_UNUSED(obj)
    
    switch (event->type()) {
        case QEvent::ToolTip:
            return true;
        case QEvent::Hide:
        case QEvent::Leave:
            hideToolTip();
            break;
        case QEvent::Enter:
            m_isEnter = true;
            if (QWidget *parent = qobject_cast<QWidget*>(this->parent())) {
                if (!parent->toolTip().isEmpty() && parent->isEnabled()) {
                    int t = parent->toolTipDuration() > 0 ? parent->toolTipDuration() : -1;
                    m_toolTip->setDuration(t);
                    m_timer->start(m_showDelay);
                }
            }
            break;
        case QEvent::MouseButtonPress:
            hideToolTip();
            break;
        default:
            break;
    }
    
    return QObject::eventFilter(obj, event);
}


void ToolTipEventFilter::showToolTip() {
    if (!m_isEnter) return;
    
    if (QWidget *parent = qobject_cast<QWidget*>(this->parent())) {
        m_toolTip->setText(parent->toolTip());
        m_toolTip->adjustPos(parent, m_position);
        m_toolTip->show();
    }
}


void ToolTipEventFilter::hideToolTip() {
    m_isEnter = false;
    m_timer->stop();
    if (m_toolTip) {
        m_toolTip->hide();
    }
}


void ToolTipEventFilter::setToolTipDelay(int delay) {
    m_showDelay = delay;
}

File: python/QEasyWidgets/Components/ToolTip.py
from typing import Optional, overload
from PyEasyUtils import singledispatchmethod
from PySide6.QtCore import *
from PySide6.QtGui import *
from PySide6.QtWidgets import *

from ..Common.Config import Position
from ..Common.StyleSheet import StyleSheetBase
from ..Common.QFunctions import getCurrentScreen, getScreenGeometry, setDropShadowEffect, setOpacityEffect

##############################################################################################################################

class ToolTipBase(QFrame):
    """
    Base class for toolTip components
    """
    def __init__(self, parent: QWidget = ..., text = '',):
        super().__init__(parent)

        self._text = text

        self._duration = 333

        self.timer = QTimer(self)
        self.timer.setSingleShot(True)
        self.timer.timeout.connect(self.hide)

        # set layout
        self.setLayout(QHBoxLayout())
        self.layout().setContentsMargins(12, 8, 12, 12)

        # set container
        self.container = self._createContainer()
        self.containerLayout = QHBoxLayout(self.container)
        self.layout().addWidget(self.container)
        self.label = QLabel(text, self)
        self.containerLayout.addWidget(self.label)
        self.containerLayout.setContentsMargins(8, 6, 8, 6)

        # add shadow
        setDropShadowEffect(self,
            blurRadius = 21,
            color = QColor(0, 0, 0, 60),
            offset_dx = 0, offset_dy = 3,
        )

        # add opacity effect
        self.opacityAnim = setOpacityEffect(self,)

        # set style
        self.setAttribute(Qt.WA_TransparentForMouseEvents)
        self.setAttribute(Qt.WA_TranslucentBackground)
        self.setWindowFlags(Qt.Tool | Qt.FramelessWindowHint)
        self._setQSS()

    def _createContainer(self):
        return QFrame(self)

    def _setQSS(self):
        self.container.setObjectName("container")
        self.label.setObjectName("contentLabel")
        StyleSheetBase.ToolTip.apply(self)
        self.label.adjustSize()
        self.adjustSize()

    def showEvent(self, e):
        self.opacityAnim.setStartValue(0)
        self.opacityAnim.setEndValue(1)
        self.opacityAnim.start()
        self.timer.stop()
        if self.duration() > 0:
            self.timer.start(self._duration + self.opacityAnim.duration())
        super().showEvent(e)

    def text(self):
        return self._text

    def setText(self, text):
        self._text = text
        self.label.setText(text)
        self.container.adjustSize()
        self.adjustSize()

    @singledispatchmethod
    def showText(self, pos: QPoint, text: str):
        self.move(pos)
        self.setText(text)
        self.show()

    @showText.register
    def _(self, x: int, y: int, text: str):
        self.move(x, y)
        self.show(text)

    def hideText(self):
        self.hide()

    def hideEvent(self, e):
        self.timer.stop()
        super().hideEvent(e)

    def duration(self):
        return self._duration

    def setDuration(self, duration: int):
        self._duration = duration

    def adjustPos(self, widget: QWidget, position: Position):
        if position == Position.Top:
            pos = widget.mapToGlobal(QPoint())
            x = pos.x() + widget.width()//2 - self.width()//2
            y = pos.y() - self.height()
        elif position == Position.Bottom:
            pos = widget.mapToGlobal(QPoint())
            x = pos.x() + widget.width()//2 - self.width()//2
            y = pos.y() + widget.height()
        elif position == Position.Left:
            pos = widget.mapToGlobal(QPoint())
            x = pos.x() - self.width()
            y = pos.y() + (widget.height() - self.height()) // 2
        elif position == Position.Right:
            pos = widget.mapToGlobal(QPoint())
            x = pos.x() + widget.width()
            y = pos.y() + (widget.height() - self.height()) // 2
        elif position == Position.TopLeft:
            pos = widget.mapToGlobal(QPoint())
            x = pos.x() - self.layout().contentsMargins().left()
            y = pos.y() - self.height()
        elif position == Position.TopRight:
            pos = widget.mapToGlobal(QPoint())
            x = pos.x() + widget.width() - self.width() + self.layout().contentsMargins().right()
            y = pos.y() - self.height()
        elif position == Position.BottomLeft:
            pos = widget.mapToGlobal(QPoint())
            x = pos.x() - self.layout().contentsMargins().left()
            y = pos.y() + widget.height()
        elif position == Position.BottomRight:
            pos = widget.mapToGlobal(QPoint())
            x = pos.x() + widget.width() - self.width() + self.layout().contentsMargins().right()
            y = pos.y() + widget.height()
        rect = getScreenGeometry(getCurrentScreen())
        x = max(rect.left(), min(pos.x(), rect.right() - self.width() - 4))
        y = max(rect.top(), min(pos.y(), rect.bottom() - self.height() - 4))
        self.move(x, y)


class ToolTipEventFilter(QObject):
    """    
    """
    def __init(self, parent: Optional[QWidget] = None, delay: int = 333):
        super().__init__(parent)

        self._showDelay = delay

        self.isEnter = False

        self.timer = QTimer(self)
        self.timer.setSingleShot(True)
        self.timer.timeout.connect(self.showToolTip)

    @singledispatchmethod
    def __init__(self, parent: Optional[QWidget] = None, position = Position.Top, delay: int = 333):
        self.__init(parent, delay)

        self.position = position

        self._toolTip = ToolTipBase(self.parent().window(), self.parent().toolTip())

    @__init__.register
    def _(self, parent: Optional[QWidget] = None, toolTip: ToolTipBase = ..., delay: int = 333):
        self.__init(parent, delay)

        self._toolTip = toolTip if isinstance(toolTip, ToolTipBase) else ToolTipBase(self.parent().window(), self.parent().toolTip())

    def eventFilter(self, obj: QObject, e: QEvent) -> bool:
        if e.type() == QEvent.ToolTip:
            return True
        elif e.type() in [QEvent.Hide, QEvent.Leave]:
            self.hideToolTip()
        elif e.type() == QEvent.Enter:
            self.isEnter = True
            parent = self.parent()
            if parent.isWidgetType() and parent.toolTip() and parent.isEnabled():
                t = parent.toolTipDuration() if parent.toolTipDuration() > 0 else -1
                self._toolTip.setDuration(t)
                self.timer.start(self._showDelay) # show toolTip after delay
        elif e.type() == QEvent.MouseButtonPress:
            self.hideToolTip()
        return super().eventFilter(obj, e)

    def hideToolTip(self):
        self.isEnter = False
        self.timer.stop()
        if self._toolTip:
            self._toolTip.hide()

    def showToolTip(self):
        if not self.isEnter:
            return

        parent = self.parent()
        self._toolTip.setText(parent.toolTip())
        self._toolTip.adjustPos(parent, self.position)
        self._toolTip.show()

    def setToolTipDelay(self, delay: int):
        self._showDelay = delay
    
##############################################################################################################################